### ISSUE TITLE
Regex not parsing as expected when I have digits in front of placeholder

### ISSUE THREAD

**Issue by ashish10alex** [sqlfluff/sqlfluff#5595]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Regex not parsing as expected when I have digits in in the front 

for example in this sql string format as expected 

```sql
SELECT
    1 AS id,
    2 AS qty, 3 AS foo
FROM ${ref("some_table_here")}
WHERE id = 1
```

However the following does not 

```sql
SELECT
    1 AS id,
    2 AS qty, 3 AS foo
FROM ${ref("10_some_table_here")}
WHERE id = 1
```

Only difference being the number 10 in the "" string 

Note that the regex used passes \d* to match a digit if there at the beginning. 

```bash
param_regex = \${ref\(\"(?P<param_name>\d*[\w]+)"\)\}
```

### Expected Behaviour

Should format the sql file 

### Observed Behaviour

Fails with   [1 unfixable linting violations found]

![CleanShot 2024-02-01 at 19 49 40@2x](https://github.com/sqlfluff/sqlfluff/assets/34306898/05188d1c-475c-435b-9272-c24ea1f257c2)

### How to reproduce

Use the following sql file 

```sql
SELECT
    1 AS id,
    2 AS qty, 3 AS foo
FROM ${ref("10_some_table_here")}
WHERE id = 1
```

### Dialect

bigquery

### Version

sqlfluff, version 2.3.5

### Configuration

```bash
[sqlfluff]
templater = placeholder
dialect = bigquery

[sqlfluff:templater:placeholder]
param_regex = \${ref\(\"(?P<param_name>\d*[\w]+)"\)\}
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
